### DISCUSSION TITLE
Debugger Visualizer

### DISCUSSION THREAD

**Discussion by mcownie** [NRules/NRules#262] [General]
just found this! its so cool :)

**Comment by snikolayev** [comment]
I'm glad you liked it. Just in case, the docs for the debugger visualizer are here: https://nrules.net/articles/diagnostics.html#debugger-visualizer

**Comment by marked23** [comment]
I had trouble getting the Debugger Visualizer working... the button was not appearing in the watch window.   What finally made it work for me, was when I ran Visual Studio as administrator.  Is running Visual Studio as administrator a requirement for the visualizer?

**Reply by snikolayev** [reply]
This is more of a question about the debugger visualizers in general. There is certainly nothing in this visualizer specifically that would require admin rights. Looking at Microsoft docs I see nothing that would imply it’s needed ([Link](https://docs.microsoft.com/en-us/visualstudio/debugger/how-to-install-a-visualizer?view=vs-2019) ). To be honest I always run visual studio elevated, so I can’t confirm or deny this.
In the docs there are two ways to install the visualizers, first is the one I described in NRules docs, to put it under My Documents. Another one is to install under visual studio install location. Not sure which approach you used, perhaps you can try the other one and see if that still requires admin rights.

**Reply by jackfobel** [reply]
Trying to get the visualizer going. I have copied the needed files into the documents > visual studio > visualizers folder.

Verified all version of assets are 1.0.2. I pulled down the repo and am using the main branch for 1.0.2

Getting this error I can't get around:
Unable to load the custom visualizer.
The system cannot find the file specified. (Exception from HRESULT: 0x80070002)

Tried the visual studio installation location to no avail as well.

Let me know if you or anyone else is able to reproduce this issue.

**Reply by snikolayev** [reply]
@jackfobel I just cloned clean main, built visualizer, installed and tried using it - it all worked.
I suggest you try do a fresh clone, clean nuget local caches, then build Core and Visualizer and re-install into Documents location (C:\Users\username\Documents\Visual Studio 2022\Visualizers).
If it still fails after this, can you please attach the screenshot of the error, and expand the stacktrace portion?
If visualizer does not work, you can also generate the DGML file from code. Assuming you have NRules session called "session", this is really what visualizer does behind the scenes:
```c#
var writer = new NRules.Diagnostics.Dgml.DgmlWriter(session.GetSchema());
writer.WriteAllText("test.dgml");
```

**Reply by jackfobel** [reply]
I have it working from a different machine now. Will investigate from the other machine why it is failing.

Thanks for the quick reply though!

**Comment by snikolayev** [comment]
FYI, there is now a debugger visualizer available as a VS Code extension: https://marketplace.visualstudio.com/items?itemName=nrules.vscode-nrules-visualizer
